pandas/core/frame.py

```diff
@@ -2160,6 +2160,150 @@ def _getitem_frame(self, key):
+    #
+    # A collection of helper methods for DataFrame operations that accept a
+    # combination of columns and index levels.  All such operations should
+    # utilize/extend these methods when possible so that we have consistent
+    # precedence and validation logic throughout the library.
+    #
+    # General Notes:
+    #
+    #   - If a column and index level share the same name, the column takes
+    #     precedence. Currently a ``FutureWarning`` should be issued in this
+    #     situation. In a future version we will convert this into an
+    #     exception.
+    #
+    #   - These methods assume axis=1
+    #
+    #   - Only string keys may be used to reference index levels.
+    #
+
+    def _get_column_or_level_values(self, key):
+        """
+        Return an array of values from a DataFrame column or named index level
+
+        Parameters
+        ----------
+        key: str or object
+            Label of column or index level. If `key` is present in the frame as
+            a column label, the corresponding column is chosen. Otherwise,
+            if `key` is a string and the is present in the frame as the name
+            of an index level, the corresponding index level is chosen.
+            Otherwise, a ``KeyError`` is raised.
+
+        Returns
+        -------
+        values: np.ndarray
+
+        Raises
+        ------
+        KeyError
+            if `key` matches neither a column label nor an index level name
+        """
+        if key in self:
+            self._check_column_or_level_ambiguity(key)
+            values = self[key]._values
+        elif self._is_index_reference(key):
+            values = self.index.get_level_values(key)._values
+        else:
+            raise KeyError(key)
+
+        return values
+
+    def _check_column_or_level_ambiguity(self, key):
+        """
+        Check whether `key` matches both a column label and an index level
+        and issue a ``FutureWarning`` if this is the case.
+
+        Note: This method will be altered to raise an ambiguity exception in
+        a future version.
+
+        Parameters
+        ----------
+        key: str or object
+            Label of column or index level
+
+        Returns
+        -------
+        ambiguous: bool
+
+        Raises
+        ------
+        FutureWarning
+            if `key` is ambiguous. This will become an ambiguity error in a
+            future version
+
+        """
+        if (isinstance(key, compat.string_types) and
+                key in self.columns and
+                key in self.index.names):
+
+            warnings.warn(
+                ("'{key}' is both a column name and an index level.\n"
+                 "Defaulting to column, but this will raise an "
+                 "ambiguity error in a future version"
+                 ).format(key=key), FutureWarning)
+
+            return True
+        else:
+            return False
+
+    def _is_index_reference(self, key):
+        """
```

Comment:
move this to generic.py it should take an axis argument and only look at that axis.

Reply:
pls add some specific tests for this

Reply:
I moved all of the new helper methods to generic.py and added axis args.
Tests for all of the new methods added in ``tests/indexes/test_label_or_level_utils.py``